### PR TITLE
Improve performance of ParticleEngineClient

### PR DESCRIPTION
**Pull request by howmanysmall**
- Prefer `WorldRoot:Raycast` to `WorldRoot:FindPartOnRay`.
- Use `table.create` where possible.
- Remove cached functions
- Stop using `tick` to calculate DeltaTime and instead use the passed DeltaTime from the Heartbeat function.
- Localize constant userdata.
- Fixed the OnClientEvent function.

### CODE REVIEW

**Comment by Jjx003** [comment]
Why replace the cached functions?

**Comment by howmanysmall** [comment]
It's not beneficial in Luau at all. It's just less readable.

**Comment by Jjx003** [comment]
I didn't realize Luau implemented these optimizations. Where can I find the details of what's been improved with Luau?

**Comment by howmanysmall** [comment]
I can probably answer some questions, but [this](https://gist.github.com/zeux/bb646a63c02ff2828117092036d2d174) as well as the DevForum are good resources.

File: Modules/Client/Particles/ParticleEngineClient.lua [hunk 1]
```diff
@@ -38,21 +31,21 @@ function ParticleEngineClient:Init(screen)
...
 	self._particles = {}
-	self._particleFrames = {}
+	self._particleFrames = table.create(self._maxParticles)
 
 	for i=1, self._maxParticles do
 		self._particleFrames[i] = newFrame("_particle")
 	end
 
-	RunService.Heartbeat:Connect(function()
+	RunService.Heartbeat:Connect(function(dt)
```
[the remark this answers is not in the document]
Comment (howmanysmall) [reply Modules/Client/Particles/ParticleEngineClient.lua:46]: Ah, I see. Didn't realize that would happen.

**Comment by Quenty** [comment]
We have two luacheck issues, I'm fixing one of them by updating the luacheck spec.

image

**Comment by Quenty** [comment]
This is all that remains: 

image

**Pull request closed without merging**